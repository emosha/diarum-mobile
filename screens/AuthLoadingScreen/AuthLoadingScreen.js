import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { AsyncStorage } from 'react-native';
import { Font } from 'expo';

import Loader from '../../components/Loader/Loader';

export default class AuthLoadingScreen extends React.Component {
  static propTypes = {
    navigation: PropTypes.object.isRequired,
  }

  async componentDidMount() {
    await this.bootstrapApp();
  }

  componentDidCatch() {
    this.props.navigation.navigate('Welcome');
  }

  /**
   * Bootstraps application
   *
   * @returns {void}
   */
  bootstrapApp = async () => {
    await this.loadFonts();
    const screen = await this.getNextScreen();

    this.props.navigation.navigate(screen);
  }

  /**
   * Get next screen
   *
   * @returns {string} next screen
   */
  getNextScreen = async () => {
    const alreadyOpened = await this.alreadyOpened();
    const loggedInUser = await this.checkUserAuthStatus();
    const unAuthScreen = alreadyOpened ? 'Signup' : 'Welcome';

    return loggedInUser ? 'App' : unAuthScreen;
  }

  /**
   * Checks user authentication status
   *
   * @returns {Object} user
   */
  checkUserAuthStatus = async () => {
    const userString = await AsyncStorage.getItem('@DIARUM:USER');
    const user = JSON.parse(userString);

    return user;
  }

  /**
   * Checks if app was previously opened
   *
   * @returns {boolean} alreadyOpened
   */
  async alreadyOpened() {
    const value = await AsyncStorage.getItem('@DIARUM:ALREADYOPEN');

    if (!value) {
      await AsyncStorage.setItem('@DIARUM:ALREADYOPEN', 'true');
      return false;
    }

    return true;
  }

  /**
   * Load app fonts
   *
   * @returns {void}
   */
  async loadFonts() {
    await Font.loadAsync({
      'open-sans-light': require('../../assets/fonts/Open_Sans/OpenSans-Light.ttf'),
      'open-sans-italic': require('../../assets/fonts/Open_Sans/OpenSans-Italic.ttf'),
      'open-sans-regular': require('../../assets/fonts/Open_Sans/OpenSans-Regular.ttf'),
      'open-sans-semi-bold': require('../../assets/fonts/Open_Sans/OpenSans-SemiBold.ttf'),
      'open-sans-bold': require('../../assets/fonts/Open_Sans/OpenSans-Bold.ttf'),
      'open-sans-extra-bold': require('../../assets/fonts/Open_Sans/OpenSans-ExtraBold.ttf'),
    });
  }

  render() {
    return (
      <Fragment>
        <Loader size="large" color="#5D59CB" />
      </Fragment>
    );
  }
}
