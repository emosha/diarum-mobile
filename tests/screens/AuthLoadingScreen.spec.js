import React from 'react';
import { AsyncStorage } from 'react-native';

import mocks from '../helpers/mocks';
import AuthLoadingScreen from '../../screens/AuthLoadingScreen/AuthLoadingScreen';

jest.mock('expo', () => ({
  Font: {
    loadAsync() {
      return null;
    },
  },
}));

describe('AuthLoadingScreen', () => {
  let Component;
  let instance;
  const getProps = navigate => ({
    navigation: {
      navigate,
    },
  });

  beforeAll(() => {
    const props = getProps(jest.fn());
    Component = mount(
      <AuthLoadingScreen navigation={props.navigation} />,
    );

    instance = Component.instance();
  });

  it('renders correctly', () => {
    expect(Component).toMatchSnapshot();
  });

  it('navigates to the welcome screen if App has never been open', async () => {
    await instance.bootstrapApp();

    expect(instance.props.navigation.navigate).toHaveBeenCalledWith('Welcome');
  });

  it('navigates to the signup screen if App has been open but user is not authenticated', async () => {
    await AsyncStorage.setItem('@DIARUM:ALREADYOPEN', 'true');

    const props = getProps(jest.fn());
    const newInstance = await mount(
      <AuthLoadingScreen navigation={props.navigation} />,
    ).instance();

    await newInstance.bootstrapApp();

    expect(newInstance.props.navigation.navigate).toHaveBeenCalledWith('Signup');
  });

  it('navigates to the Home page if App has been open and user is authenticated', async () => {
    const props = getProps(jest.fn());
    const { createUser: { user } } = mocks[0].result.data;

    await AsyncStorage.setItem('@DIARUM:ALREADYOPEN', 'true');
    await AsyncStorage.setItem('@DIARUM:USER', JSON.stringify(user));

    const newInstance = await mount(
      <AuthLoadingScreen navigation={props.navigation} />,
    ).instance();

    await newInstance.bootstrapApp();

    expect(newInstance.props.navigation.navigate).toHaveBeenCalledWith('App');
  });

  it('navigates to the welcome screen if an error occurs', async () => {
    await instance.componentDidCatch();

    expect(instance.props.navigation.navigate).toHaveBeenCalledWith('Welcome');
  });
});
