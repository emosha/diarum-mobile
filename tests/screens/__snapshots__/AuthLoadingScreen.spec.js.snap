// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthLoadingScreen renders correctly 1`] = `
<AuthLoadingScreen
  navigation={
    Object {
      "navigate": [MockFunction],
    }
  }
>
  <Loader
    color="#5D59CB"
    size="large"
  >
    <Component
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
            "padding": 10,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "flexDirection": "row",
              "justifyContent": "space-around",
              "padding": 10,
            },
          ]
        }
      >
        <Component
          animating={true}
          color="#5D59CB"
          hidesWhenStopped={true}
          size="large"
        >
          <ActivityIndicator
            animating={true}
            color="#5D59CB"
            hidesWhenStopped={true}
            size="large"
          />
        </Component>
      </View>
    </Component>
  </Loader>
</AuthLoadingScreen>
`;
